EDUCATOR-1227 Assets Search UI

thallada opened the pull request:
- Adds AssetsSearch
- Adds AssetsResultCount
- Adds AssetsClearFiltersButton
- Restructures AssetsPage DOM to include new components
- Translates the strings in AssetsPage
- Adds new SEARCH actions/reducers
- Updates the clearFilters action so that it also clears the current search
- Conditionally show the new UI with studioContext setting (thanks @dsjen!)

I decided to pull the status alert refactor out into [a separate PR](https://github.com/edx/studio-frontend/compare/thallada/search...thallada/refactor-status-alert) since it was getting pretty gnarly.

~~I'll update with a comment once a sandbox for this PR is created.~~ I need to merge this PR and then make a separate one in platform and then I will be able to make a sandbox.

@edx/educator-dahlia

MichaelRoytman commented:
If you want, couldn't you install studio-frontend into edx-platform using a Git hash just to get a sandbox up?

MichaelRoytman commented on src/components/AssetsClearFiltersButton/index.jsx:
```diff
@@ -0,0 +1,33 @@
+import messages from './displayMessages';
+
+
+const AssetsClearFiltersButton = ({ clearFilters, courseDetails }) => (
```
Nit: If this button clears both the filters and the search now, I wonder if we can find a better name for it. I can't really think of a good one, so if no one else is opposed, it's fine. Maybe, `AssetsShowAllFilesButton`? Although, like Farhanah mentioned, I could see that being confusing too.

dsjen commented on src/components/AssetsSearch/index.jsx:
```diff
@@ -0,0 +1,89 @@
+  }
+
+  render() {
+    // TODO: InputText creates it's own form-group div. Nesting them is not semantic.
```
Nit: spelling on "it's" :)

thallada replied:
Ugh, I always get that wrong.

dsjen commented on src/components/AssetsSearch/AssetsSearch.scss:
```diff
@@ -0,0 +1,9 @@
+.form-inline label {
+  margin-right: 10px;
+  font-weight: 600;
```
Can we use [$font-weight-bold]($font-weight-bold) here? Doesn't everyone love theming?

dsjen replied:
I guess alternatively one could extend the `font-weight-bold` css class too--seems like overkill perhaps though.

thallada replied:
I tried that but it is set to 700 and I think it looks too "heavy". What do you think?
![screenshot-20180226174240-472x84](https://user-images.githubusercontent.com/1505923/36700233-185c6676-1b1d-11e8-9232-72afe6cd3a13.png)

vs. what it is set to now (600):
![screenshot-20180226174820-461x89](https://user-images.githubusercontent.com/1505923/36700292-40c389a0-1b1d-11e8-97eb-c7ba30405267.png)

dsjen replied:
Yeah, I see what you mean! @abbyluke, do you have an opinion on this? Is the [edx-bootstrap bold](https://github.com/edx/edx-bootstrap/blob/f0a6237174de7dd65815b07c73da127f082faa4d/sass/edx/_variables.scss#L237) in this case too bold?

dsjen commented on src/components/AssetsSearch/AssetsSearch.scss:
```diff
@@ -0,0 +1,9 @@
+.form-inline label {
+  margin-right: 10px;
```
I think this could be `$spacer` just in case someone wants to theme spacing, sort of like what we do in https://github.com/edx/studio-frontend/blob/master/src/components/AccessibilityPolicyForm/AccessibilityPolicyForm.scss#L4

thallada replied:
So `$spacer` ends up being 16px because of the weird `base-rem-size` Studio has. I could do `1rem` instead for 10px and `0.8rem` for 8px. It would still be kinda configurable if they changed the `base-rem-size` (everything would scale with that).

dsjen commented on src/components/AssetsResultsCount/AssetsResultsCount.scss:
```diff
@@ -0,0 +1,3 @@
+.result-count-wrapper {
+  padding: 10px;
```
Maybe use [$spacer](https://github.com/edx/edx-bootstrap/blob/master/sass/open-edx/_variables.scss#L120)?

dsjen commented on src/utils/getAssetsFilters.jsx:
```diff
@@ -31,3 +31,7 @@ export const getFilterStateObjectFromArray = (enabledFilters) => {
 };
+
+export const hasSearchOrFilterApplied = (filters, search) => (
+  Boolean(hasSelectedFilters(filters) || (search && search.length > 0))
```
Out of curiosity, why cast it to a `Boolean`?

thallada replied:
If `search` is set to something falsey, the expression will just return the value. Wrapping it in this cast will make sure it only ever returns `true` or `false`.

dsjen replied:
Gotcha. Thanks!

dsjen commented on src/utils/getAssetsFilters.test.jsx:
```diff
@@ -77,4 +77,37 @@ describe('getAssetsFilters Utility Functions', () => {
       expect(filterStateObject).toEqual(expectedfilterStateObject);
     });
   });
+  describe('hasSearchOrFilterApplied', () => {
```
Nice testing!

dsjen commented on src/components/AssetsPage/index.jsx:
```diff
@@ -88,17 +88,40 @@ export default class AssetsPage extends React.Component {
+                <WrappedAssetsResultsCount />
+              </div>
+              <div className={`${edxBootstrap['col-md-4']} ${edxBootstrap['text-right']}`}>
+                {hasSearchOrFilterApplied(this.props.filtersMetaData.assetTypes,
```
Thanks for covering this use case!

thallada commented:
@MichaelRoytman I think I finally found a way to do what you mentioned. The key was changing our `prepublishOnly` script that runs "npm run build" to `prepare` instead, which will run when studio-frontend is installed as a github URL, but not when installed from the NPM registry.

> `npm install <git remote url>`:
...
If the package being installed contains a prepare script, its dependencies and devDependencies will be installed, and the prepare script will be run, before the package is packaged and installed.
-- https://docs.npmjs.com/cli/install

So, when the sandbox runs `npm install` it will pull down this branch and then run the prepare script which will run the webpack build and emit the /dist files that are needed for the later static asset steps. I'm starting a sandbox build now.

Unrelated to all of that, but I noticed that `npm run build` was failing on this branch (seemed to be caused by paragon 2.1.0) with a weird uglifyjs error, but I fixed it by upgrading the uglifyjs plugin :man_shrugging: . The version included with webpack 3 is pretty old.

A side-effect of changing to the `prepare` script is that Travis will now run `npm run build`. It will slow our builds down by like 35 seconds but I think it is worth it so we catch errors with the production build early.

efischer19 commented:
A thumb on the comment feels insufficient here - great find @thallada!

efischer19 commented on src/components/AssetsResultsCount/displayMessages.jsx:
```diff
@@ -0,0 +1,16 @@
+const messages = defineMessages({
+  assetsResultsCountFiltered: {
+    id: 'assetsResultsCountFiltered',
+    defaultMessage: 'Showing {start}-{end} out of {total} possible matches.',
```
I'm curious to see how this gets represented as it goes up to and back from Transifex.

efischer19 replied:
I've filed https://openedx.atlassian.net/browse/EDUCATOR-2438 to track the needed i18n work for pluralization, go ahead and merge these messages as they are for now. We'll have to deal with the possibility of seeing `1 possible matches` until that gets resolved.

efischer19 commented on src/components/AssetsResultsCount/index.jsx:
```diff
@@ -0,0 +1,53 @@
+  <span className={`${edxBootstrap['font-weight-bold']}`}>
+    <FormattedNumber value={count} />
+  </span>
+);
```
This function is super neat

🆒

efischer19 commented on src/components/AssetsResultsCount/index.jsx:
```diff
@@ -0,0 +1,53 @@
+        values={{
+          start: renderCount(paginationMetaData.start + 1),
+          end: renderCount(paginationMetaData.end),
+          total: renderCount(paginationMetaData.totalCount),
```
I'm not sure if using `renderCount` like this will work correctly with the pluralization I mentioned above, I think that needs a plain number. Surely there's something for this scenario...

LMK when you get to this portion, I'll help get it ironed out if it's tricky. I'm also okay with ignoring the pluralization work in the short term (still to be done later) if it's going to delay this large PR from landing.

thallada replied:
Maybe [`<FormattedPlural>`](https://github.com/yahoo/react-intl/wiki/Components#formattedplural) could do it? It looks like it allows embedding the text inside elements (just like `<FormattedMessage>`). Since this means we will have to make separate displayMessages for "file"/"total files" and "possible match"/"possible matches" strings, we would need to make a `<WrappedFormattedPlural>` so each plural string is wrapped in the proper lang.

I still don't get how translators would specify other plural type translations that English doesn't need (zero, few, many, etc.) and how our code would know to use them. Do we just define a separate displayMessage for each plural type and just use the same word in English for most of the defaults?

efischer19 commented on src/data/i18n/default/src/components/AssetsClearFiltersButton/displayMessages.json:
```diff
@@ -0,0 +1,7 @@
+    "description": "Label for a button that clears all filters applied to a table.",
+    "defaultMessage": "Clear all filters"
+  }
+]
```
These are tool-generated; all `<component>/displayMessages.json files under `src/data/i18n/default` will be this way.

MichaelRoytman replied:
Ah, I see. Thanks for clarifying.

thallada commented on src/components/AssetsPage/container.jsx:
```diff
@@ -10,10 +10,11 @@ const mapStateToProps = state => ({
   uploadSettings: state.studioDetails.upload_settings,
   status: state.metadata.status,
   filtersMetaData: state.metadata.filters,
+  searchMetaData: state.metadata.search,
```
I agree. I like `Metadata` over `MetaData`. I can just do a find and replace on that.

thallada commented on src/components/AssetsPage/index.jsx:
```diff
@@ -88,17 +88,40 @@ export default class AssetsPage extends React.Component {
-      <h3>0 files in your course</h3>
-      <h4>Enhance your course content by uploading files such as images and documents.</h4>
-    </React.Fragment>
+    <div>
```
I think I might have done that on your suggestion but forgot to change it back when I realized that wasn't the issue.

thallada commented on src/components/AssetsClearFiltersButton/index.jsx:
```diff
@@ -0,0 +1,33 @@
+import messages from './displayMessages';
+
+
+const AssetsClearFiltersButton = ({ clearFilters, courseDetails }) => (
```
Yeah I'm not sure where to go with this. To me a search is something you use to filter, but we have also distinguished filters as something different than search in the code. I like calling it a "clear" because that's the action it performs, but what is it clearing? `AssetsClearFiltersAndSearchButton`, `AssetsClearControlsButton`, `AssetsClearQueryButton`...?

thallada commented on src/components/AssetsClearFiltersButton/displayMessages.jsx:
```diff
@@ -0,0 +1,11 @@
+const messages = defineMessages({
+  assetsClearFiltersButtonLabel: {
+    id: 'assetsClearFiltersButtonLabel',
+    defaultMessage: 'Clear all filters',
```
I think Sylvia had concerns over the wording when it was always displayed. Maybe it's not an issue when it's hidden while no search or filters are applied. Whatever we decide for the variable name I think we should use here too.

thallada commented on src/data/i18n/default/src/components/AssetsPage/displayMessages.json:
```diff
@@ -0,0 +1,27 @@
+[
+  {
+    "id": "assetsPageNoResultsNumFiles",
```
Yeah, I should be more consistent. `Count` works for me :+1:

MichaelRoytman commented:
I just noticed this. When I'm viewing a course with no assets, the search bar also shows up. Since we're hiding all other filters, I think we should probably hide the search bar, especially since there is nothing to search for.

![image](https://user-images.githubusercontent.com/11871801/36905086-66c9748a-1e00-11e8-91a1-8585619b6d40.png)

thallada commented on src/components/AssetsPage/displayMessages.jsx:
```diff
@@ -0,0 +1,31 @@
+    defaultMessage: 'No files were found.',
+    description: 'Message displayed in table when no results are found.',
+  },
+  assetsPageNoResultsClear: {
```
Oh yes, good catch!

thallada commented:
@MichaelRoytman good point. The search bar will now only display on a "NORMAL" page (has assets and not filtered down to 0).

I think I addressed all of the comments except for the point about the pluralization which we decided to address in a separate PR later. Let me know if there is anything else, otherwise I think this is good to merge now!

efischer19 commented on Makefile:
```diff
@@ -65,3 +65,6 @@ pull_translations: ## must be exactly this name for edx tooling support, see eco
 	git diff --name-only --exit-code package-lock.json
+
+copy-dist:
+	for f in dist/*; do docker cp $$f edx.devstack.studio:/edx/app/edxapp/edx-platform/node_modules/@edx/studio-frontend/dist/; done
```
🙌
